RestTemplate for eureka client failed to send heartbeat after eureka restart

If in EDGWARE i remove jersey the possibility introduced to use RestTemplate has a different behaviour.   A client fails to send heart beat to eureka server after eureka restart. 
I think it depends on InstanceInfo or something linked with charsets

I also see this issue with Spring Boot 2.0.1.RELEASE and Spring Cloud Netflix 2.0.0.RC1.  It only happens when I am not using the jersey1 eureka client.

```
[2018-05-07 19:23:37:672] [WARN] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient] Request execution failed with message: 404 null
[2018-05-07 19:23:37:704] [ERROR] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient] Request execution error
org.springframework.web.client.HttpClientErrorException: 404 null
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:94)
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:79)
	at org.springframework.web.client.ResponseErrorHandler.handleError(ResponseErrorHandler.java:63)
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:775)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:728)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:684)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:600)
	at org.springframework.cloud.netflix.eureka.http.RestTemplateEurekaHttpClient.sendHeartBeat(RestTemplateEurekaHttpClient.java:98)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$3.execute(EurekaHttpClientDecorator.java:92)
	at com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient.executeOnNewServer(RedirectingEurekaHttpClient.java:118)
	at com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient.execute(RedirectingEurekaHttpClient.java:79)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$3.execute(EurekaHttpClientDecorator.java:92)
	at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:120)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$3.execute(EurekaHttpClientDecorator.java:92)
	at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.DiscoveryClient.renew(DiscoveryClient.java:846)
	at com.netflix.discovery.DiscoveryClient$HeartbeatThread.run(DiscoveryClient.java:1399)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[2018-05-07 19:23:37:704] [WARN] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient] Request execution failed with message: 404 null
[2018-05-07 19:23:37:704] [ERROR] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.DiscoveryClient] DiscoveryClient_NOT-YOURSERVICE/NOT-YOURSERVICE-af442b6e1d14 - was unable to send heartbeat!
com.netflix.discovery.shared.transport.TransportException: Retry limit reached; giving up on completing the request
	at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:139)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$3.execute(EurekaHttpClientDecorator.java:92)
	at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77)
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.sendHeartBeat(EurekaHttpClientDecorator.java:89)
	at com.netflix.discovery.DiscoveryClient.renew(DiscoveryClient.java:846)
	at com.netflix.discovery.DiscoveryClient$HeartbeatThread.run(DiscoveryClient.java:1399)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

@ryanjbaxter I tracked down the issue

[RestTemplateEurekaHttpClient.sendHeartBeat](https://github.com/spring-cloud/spring-cloud-netflix/blob/master/spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateEurekaHttpClient.java#L98) needs to handle HttpClientErrorException thrown from DefaultResponseErrorHandler to get the status code and return a response with the status code.
```
[2018-06-19 11:26:10:023] [ERROR] [DiscoveryClient-HeartbeatExecutor-0] [com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient]  
 + Message: Request execution error\n+ Throwable: org.springframework.web.client.HttpClientErrorException: 404 null
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:94)
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:79)
	at org.springframework.web.client.ResponseErrorHandler.handleError(ResponseErrorHandler.java:63)
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:766)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:724)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:680)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:600)
	at org.springframework.cloud.netflix.eureka.http.RestTemplateEurekaHttpClient.sendHeartBeat(RestTemplateEurekaHttpClient.java:98)
```


I tested locally using Eureka 1.9.2 - using RestTemplate for Eureka Client vs Jersey

State 1:

- Eureka server UP
- Client application UP
- Client application registiers to Eureka Server
- client application heartbeats fine

State 2:

- kill eureka server
- client application errors trying to heartbeats
- wait a few minutes

State 3:

- Start eureka server UP
- client application errors with 404 error on heartbeat

Eureka Service logs
```
PUT /eureka/apps/MYAPP/MYAPP-myhostname?status=UP&lastDirtyTimestamp=1529425763653 HTTP/1.1" 404
[WARN] [http-nio-8080-exec-4] [com.netflix.eureka.resources.InstanceResource] Not Found (Renew): MYAPP - MYAPP-myhostname
```

--- a/spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateTransportClientFactory.java
+++ b/spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateTransportClientFactory.java
@@ -19,8 +19,10 @@ package org.springframework.cloud.netflix.eureka.http;
 import java.net.URI;
 import java.net.URISyntaxException;
 
+import org.springframework.http.HttpStatus;
 import org.springframework.http.client.support.BasicAuthorizationInterceptor;
 import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
+import org.springframework.web.client.DefaultResponseErrorHandler;
 import org.springframework.web.client.RestTemplate;
 
 import com.fasterxml.jackson.databind.BeanDescription;
@@ -74,6 +76,7 @@ public class RestTemplateTransportClientFactory implements TransportClientFactor
 		}
 
 		restTemplate.getMessageConverters().add(0, mappingJacksonHttpMessageConverter());
+		restTemplate.setErrorHandler(new ErrorHanlder());
 
 		return restTemplate;
 	}
@@ -132,6 +135,22 @@ public class RestTemplateTransportClientFactory implements TransportClientFactor
 
 	@Override
 	public void shutdown() {
+	}
+
+	class ErrorHanlder extends DefaultResponseErrorHandler {
+		@Override
+		protected boolean hasError(HttpStatus statusCode) {
+			/**
+			 * When the Eureka server restarts and a client tries to sent a heartbeat the server
+			 * will respond with a 404.  By default RestTemplate will throw an exception in this case.
+			 * What we want is to return the 404 to the upstream code so it will send another registration
+			 * request to the server.
+			 */
+			if(statusCode.is4xxClientError()) {
+				return false;
+			}
+			return super.hasError(statusCode);
+		}
 	}
 
 }

--- a/spring-cloud-netflix-eureka-client/src/test/java/org/springframework/cloud/netflix/eureka/http/EurekaServerMockApplication.java
+++ b/spring-cloud-netflix-eureka-client/src/test/java/org/springframework/cloud/netflix/eureka/http/EurekaServerMockApplication.java
@@ -20,6 +20,7 @@ import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
 import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
 import org.springframework.web.bind.annotation.DeleteMapping;
 import org.springframework.web.bind.annotation.GetMapping;
@@ -82,13 +83,16 @@ public class EurekaServerMockApplication {
 	@ResponseStatus(HttpStatus.OK)
 	@PutMapping(value = "/apps/{appName}/{id}", params = { "status",
 			"lastDirtyTimestamp" })
-	public InstanceInfo sendHeartBeat(@PathVariable String appName,
-			@PathVariable String id, @RequestParam String status,
-			@RequestParam String lastDirtyTimestamp,
-			@RequestParam(required = false) String overriddenstatus) {
-		return new InstanceInfo(null, null, null, null, null, null, null, null, null,
+	public ResponseEntity sendHeartBeat(@PathVariable String appName,
+										@PathVariable String id, @RequestParam String status,
+										@RequestParam String lastDirtyTimestamp,
+										@RequestParam(required = false) String overriddenstatus) {
+		if("fourOFour".equals(appName)) {
+			return new ResponseEntity(HttpStatus.NOT_FOUND);
+		}
+		return new ResponseEntity<InstanceInfo>(new InstanceInfo(null, null, null, null, null, null, null, null, null,
 				null, null, null, null, 0, null, null, null, null, null, null, null, 0l,
-				0l, null, null);
+				0l, null, null), HttpStatus.OK);
 	}
 
 	@ResponseStatus(HttpStatus.OK)

--- a/spring-cloud-netflix-eureka-client/src/test/java/org/springframework/cloud/netflix/eureka/http/RestTemplateEurekaHttpClientTest.java
+++ b/spring-cloud-netflix-eureka-client/src/test/java/org/springframework/cloud/netflix/eureka/http/RestTemplateEurekaHttpClientTest.java
@@ -92,6 +92,12 @@ public class RestTemplateEurekaHttpClientTest {
 	}
 
 	@Test
+	public void testSendHeartBeatFourOFour() {
+		Assert.assertEquals(HttpStatus.NOT_FOUND.value(), eurekaHttpClient
+				.sendHeartBeat("fourOFour", "test", info, null).getStatusCode());
+	}
+
+	@Test
 	public void testStatusUpdate() {
 		Assert.assertEquals(HttpStatus.OK.value(), eurekaHttpClient
 				.statusUpdate("test", "test", InstanceStatus.UP, info).getStatusCode());

